internal/terraform/context.go

```diff
@@ -337,6 +338,51 @@ func (c *Context) watchStop(walker *ContextGraphWalker) (chan struct{}, <-chan s
 	return stop, wait
 }
 
+func (c *Context) checkStateDependencies(state *states.State) tfdiags.Diagnostics {
+	if state == nil {
+		// no state is fine, first time execution etc
+		return nil
+	}
+
+	var diags tfdiags.Diagnostics
+	for providerAddr := range state.ProviderRequirements() {
+		if !c.plugins.HasProvider(providerAddr) {
+			if c.plugins.HasPreloadedSchemaForProvider(providerAddr) {
+				// If the caller provided a preloaded schema for this provider
+				// then we'll take that as a hint that the caller is intending
+				// to handle some of these pre-validation tasks itself and
+				// so we'll just optimistically assume that the caller
+				// has arranged for this to work some other way, or will
+				// return its own version of this error before calling
+				// into here if not.
+				continue
+			}
+			if !providerAddr.IsBuiltIn() {
+				diags = diags.Append(tfdiags.Sourceless(
+					tfdiags.Error,
+					"Missing required provider",
+					fmt.Sprintf(
+						"This state requires provider %s, but that provider isn't available. You may be able to install it automatically by running:\n  terraform init",
```

Comment:
> You may be able to install it automatically

What is that dependent on? The provider being in the required provider section?

Reply:
Defined anywhere in the configuration or the state. This basically catches someone running `plan` or `apply` before they ran `init`. We do the same check with a similar error message in case something is in config and hasn't been caught: https://github.com/hashicorp/terraform/blob/main/internal/terraform/context.go#L389

`init` won't install automatically if the configuration is wrong, or it's not a default provider and isn't in the required providers. But all cases `init` will then provide specific error messages for why something couldn't be installed.